include/ensmallen_bits/cmaes/weight_init_policies/negative_weight.hpp

```diff
@@ -0,0 +1,156 @@
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ * 
+ */
+#ifndef ENSMALLEN_CMAES_NEGATIVE_WEIGHT_HPP
+#define ENSMALLEN_CMAES_NEGATIVE_WEIGHT_HPP
+
+namespace ens{
+
+class NegativeWeight{
+
+ public:
+  /**
+   * Constructor
+   * 
+   */
+  NegativeWeight()
+  {
+    // Doing nothing
+  }
+  
+  /**
+   * This function will generate raw weights and mu_eff first 
+   * 
+   * @param lambda The length of raw weights 
+   */
+  void GenerateRaw(const size_t lambda)
+  {
+    // Checking the length of the weights vector
+    len = lambda;
+    assert(len >= 2 && "Number of weights must be >= 2");
+    weights = std::log((len + 1) / 2) - arma::log(arma::linspace<arma::Row<double> >(0, len - 1, len) + 1.0);
+    
+    assert(weights(0) > 0 && "The first weight must be >0");
+    assert(weights(len-1) <= 0 && "The last weight must be <= 0");
+    // mu is expected a half of len(allias of lambda in CMAparameters class)
+    mu = 0;
+    for(size_t i = 0; i < len; ++i)
+    {
+      if(weights(i) > 0) mu++;
+    }
+    double sumPos = arma::accu(weights.cols(0, mu-1));
+    // positive weights sum to one
+    for(size_t i = 0; i < mu; ++i)
+    {
+      weights(i) /= sumPos;
+    }
+    mu_eff = 1 / arma::accu(arma::pow(weights.cols(0, mu-1), 2));
+  }
+
+  /**
+   * Generate negative weight for new population
+   * 
+   *  @tparam ElemType The type of elements in weight vector
+   *  @param dim Dimension of iterate variable
+   *  @param c1 
+   *  @param cmu
+   */
+  arma::Row<double> Generate(const size_t dim,
+                               const double c1,
+                               const double cmu)
+  {
+    if(c1 > 10 * cmu)
+      std::cout << "Warning: c1/cmu seems to assume a too large value for negative weight setting" << std::endl;
+        
+    double sumNeg = std::abs(arma::accu(weights.cols(mu, len-1)));
+
+    const double alpha_mu_negative = 1 + c1/cmu;
+    const double alpha_posdef_negative = (1 - c1 - cmu) / (dim * cmu);
+    double factor = std::min(alpha_mu_negative, alpha_posdef_negative); 
+
+    for(size_t i = mu; i < len; ++i)
+    {
+      weights(i) *= factor;
+      weights(i) /= sumNeg; 
+    } 
+
+    const double alpha_mu_eff_negative = 1 + 2 * NegativeEff(weights) / (mu_eff + 2);
+    if(std::abs(arma::accu(weights.cols(mu, len-1))) >= -std::abs(alpha_mu_eff_negative))
+    {
+      factor = abs(alpha_mu_eff_negative) / std::abs(arma::accu(weights.cols(mu, len-1)));
+      if(factor < 1)
+      {
+        for(size_t i = mu; i < len; ++i)
+        {
+          weights(i) *= factor;
+        }
+      }
+    }
+    mu_eff_neg = NegativeEff(weights);
+    Checking();
+    return weights;
+  }
+  /**
+   * This function will check all the conditions that the weights vector has to stastify
+   */
+  void Checking()
```

Comment:
We should add an option to enable the extra checking, to accelerate the executation.